src/tools_for_rules.jl

```diff
@@ -0,0 +1,477 @@
+
+julia> foo_varargs(x...) = 5
+foo_varargs (generic function with 1 method)
+
+julia> @zero_adjoint DefaultCtx Tuple{typeof(foo_varargs), Vararg}
+
+julia> is_primitive(DefaultCtx, Tuple{typeof(foo_varargs), Any, Float64, Int})
+true
+
+julia> rrule!!(zero_fcodual(foo_varargs), zero_fcodual(3.0), zero_fcodual(5))[2](NoRData())
+(NoRData(), 0.0, NoRData())
+```
+Be aware that it is not currently possible to specify any of the type parameters of the
+`Vararg`. For example, the signature `Tuple{typeof(foo), Vararg{Float64, 5}}` will not work
+with this macro.
+
+WARNING: this is only correct if the output of the function does not alias any fields of the
+function, or any of its arguments. For example, applying this macro to the function `x -> x`
+will yield incorrect results.
+
+As always, you should use [`TestUtils.test_rule`](@ref) to ensure that you've not
+made a mistake.
+
+# Signatures Unsupported By This Macro
+
+If the signature you wish to apply `@zero_adjoint` to is not supported, for example because
+it uses a `Vararg` with a type parameter, you can still make use of
+[`zero_adjoint`](@ref).
+"""
+macro zero_adjoint(ctx, sig)
+
+    # Parse the signature, and construct the rule definition. If it is a vararg definition,
+    # then the last argument requires special treatment.
+    arg_type_symbols, where_params = parse_signature_expr(sig)
+    arg_names = map(n -> Symbol("x_$n"), eachindex(arg_type_symbols))
+    is_vararg = arg_type_symbols[end] === :Vararg
+    if is_vararg
+        arg_types = vcat(
+            map(t -> :(Mooncake.CoDual{<:$t}), arg_type_symbols[1:end-1]),
+            :(Vararg{Mooncake.CoDual}),
+        )
+        splat_symbol = Expr(Symbol("..."), arg_names[end])
+        body = Expr(
+            :call, Mooncake.zero_adjoint, arg_names[1:end-1]..., splat_symbol,
+        )
+    else
+        arg_types = map(t -> :(Mooncake.CoDual{<:$t}), arg_type_symbols)
+        body = Expr(:call, Mooncake.zero_adjoint, arg_names...)
+    end
+
+    # Return code to create a method of is_primitive and a rule.
+    ex = quote
+        Mooncake.is_primitive(::Type{$ctx}, ::Type{<:$sig}) = true
+        $(construct_def(arg_names, arg_types, where_params, body))
+    end
+    return esc(ex)
+end
+
+#
+# Functionality supporting @from_rrule
+#
+
+"""
+    to_cr_tangent(t)
+
+Convert a Mooncake tangent into a type that ChainRules.jl `rrule`s expect to see.
+"""
+to_cr_tangent(t::IEEEFloat) = t
+to_cr_tangent(t::Array{<:IEEEFloat}) = t
+to_cr_tangent(::NoTangent) = ChainRulesCore.NoTangent()
+
+"""
+    increment_and_get_rdata!(fdata, zero_rdata, cr_tangent)
+
+Increment `fdata` by the fdata component of the ChainRules.jl-style tangent, `cr_tangent`,
+and return the rdata component of `cr_tangent` by adding it to `zero_rdata`.
+"""
+increment_and_get_rdata!(::NoFData, r::T, t::T) where {T<:IEEEFloat} = r + t
+function increment_and_get_rdata!(f::Array{P}, ::NoRData, t::Array{P}) where {P<:IEEEFloat}
+    increment!!(f, t)
```

Comment:
I don't really know what this does, but it jumps at me as a call to `!!` that ignores the return value. If `increment!!` is guaranteed to mutate rather than create a new instance, could it be called `increment!`? If not, might this fail to do what is expected of it?

Reply:
Thanks for pointing this out. In this particular case, I know for sure that this is safe to do.